File: packages/runpane/src/installers.ts
import childProcess from 'child_process';
import fs from 'fs';
import os from 'os';
import path from 'path';
import type { ArtifactFormat, InstallTarget, ParsedArgs } from './commands';
import type { DownloadedArtifact } from './download';
import { defaultInstallRoot, type PanePlatform } from './platform';

export interface InstalledPane {
  executablePath: string;
  installKind: 'existing' | 'installed' | 'launched-installer';
}

export function resolveExistingPanePath(panePath?: string): string | undefined {
  if (panePath) {
    return fs.existsSync(panePath) ? panePath : undefined;
  }

  const candidates = [
    process.platform === 'darwin' ? '/Applications/Pane.app/Contents/MacOS/Pane' : undefined,
    process.platform === 'darwin' ? path.join(os.homedir(), 'Applications', 'Pane.app', 'Contents', 'MacOS', 'Pane') : undefined,
    process.platform === 'linux' ? path.join(os.homedir(), '.local', 'bin', 'pane') : undefined,
    process.platform === 'linux' ? '/usr/bin/pane' : undefined,
    process.platform === 'linux' ? '/opt/Pane/pane' : undefined,
    process.platform === 'linux' ? '/opt/Pane/Pane' : undefined,
    process.platform === 'win32' && process.env.LOCALAPPDATA ? path.join(process.env.LOCALAPPDATA, 'Programs', 'Pane', 'Pane.exe') : undefined,
    process.platform === 'win32' && process.env.LOCALAPPDATA ? path.join(process.env.LOCALAPPDATA, 'Pane', 'Pane.exe') : undefined,
    process.platform === 'win32' && process.env.ProgramFiles ? path.join(process.env.ProgramFiles, 'Pane', 'Pane.exe') : undefined
  ].filter((candidate): candidate is string => Boolean(candidate));

  return candidates.find((candidate) => fs.existsSync(candidate));
}

export function shouldReuseExistingPane(parsed: Pick<ParsedArgs, 'command'>, target: InstallTarget): boolean {
  return parsed.command === 'install' && target === 'daemon';
}

export async function installPaneArtifact(
  artifact: DownloadedArtifact,
  options: {
    parsed: ParsedArgs;
    platform: PanePlatform;
    format: Exclude<ArtifactFormat, 'auto'>;
    target: InstallTarget;
  }
): Promise<InstalledPane> {
  const existing = resolveExistingPanePath(options.parsed.panePath);
  if (existing && shouldReuseExistingPane(options.parsed, options.target)) {
    return { executablePath: existing, installKind: 'existing' };
  }

  if (options.platform.os === 'darwin') {
    return installMac(artifact, options);
  }
  if (options.platform.os === 'linux') {
    return installLinux(artifact, options);
  }
  return installWindows(artifact, options);
}

export function spawnPane(executablePath: string, args: string[]): Promise<number> {
  return new Promise((resolve) => {
    const child = childProcess.spawn(executablePath, buildPaneDaemonArgs(args), {
      stdio: 'inherit',
      shell: false,
      env: buildPaneDaemonEnvironment()
    });
    child.on('close', (code) => resolve(code ?? 1));
    child.on('error', (error) => {
      console.error(`Failed to launch Pane: ${error.message}`);
      resolve(1);
    });
  });
}

export interface CapturedPaneResult {
  code: number;
  stdout: string;
  stderr: string;
}

export function spawnPaneCaptured(executablePath: string, args: string[]): Promise<CapturedPaneResult> {
  return new Promise((resolve) => {
    const child = childProcess.spawn(executablePath, buildPaneDaemonArgs(args), {
      stdio: ['ignore', 'pipe', 'pipe'],
      shell: false,
      env: buildPaneDaemonEnvironment()
    });
    let stdout = '';
    let stderr = '';
    child.stdout.setEncoding('utf8');
    child.stderr.setEncoding('utf8');
    child.stdout.on('data', (chunk: string) => { stdout += chunk; });
    child.stderr.on('data', (chunk: string) => { stderr += chunk; });
    child.on('close', (code) => resolve({ code: code ?? 1, stdout, stderr }));
    child.on('error', (error) => resolve({ code: 1, stdout, stderr: `${stderr}${error.message}` }));
  });
}

// Ozone resolves its platform before the app's main script runs, so headless
// selection must arrive as argv. ELECTRON_OZONE_PLATFORM_HINT (below) was
// removed in Electron 38 and is a no-op from 39 on; it is kept only so older
// Pane builds keep working.
export function buildPaneDaemonArgs(args: string[], platform = process.platform): string[] {
  if (platform !== 'linux') {
    return [...args];
  }

  return ['--ozone-platform=headless', '--disable-gpu', ...args];
}

export function buildPaneDaemonEnvironment(
  platform = process.platform,
  baseEnvironment: NodeJS.ProcessEnv = process.env
): NodeJS.ProcessEnv {
  if (platform !== 'linux') {
    return { ...baseEnvironment };
  }

  return {
    ...baseEnvironment,
    ELECTRON_OZONE_PLATFORM_HINT: 'headless'
  };
}

export function launchPaneClient(executablePath: string): void {
  const child = childProcess.spawn(executablePath, [], {
    detached: true,
    stdio: 'ignore'
  });
  child.unref();
}

function installMac(
  artifact: DownloadedArtifact,
  { format, target }: { format: Exclude<ArtifactFormat, 'auto'>; target: InstallTarget }
): InstalledPane {
  if (format === 'dmg') {
    childProcess.spawnSync('open', [artifact.path], { stdio: 'inherit' });
    return { executablePath: '/Applications/Pane.app/Contents/MacOS/Pane', installKind: 'launched-installer' };
  }

  const appsRoot = path.join(os.homedir(), 'Applications');
  const appPath = path.join(appsRoot, 'Pane.app');
  fs.mkdirSync(appsRoot, { recursive: true });
  childProcess.execFileSync('ditto', ['-x', '-k', artifact.path, appsRoot], { stdio: 'inherit' });
  const executablePath = path.join(appPath, 'Contents', 'MacOS', 'Pane');

  if (!fs.existsSync(executablePath)) {
    throw new Error(`Pane executable was not found after extracting ${artifact.fileName}. Expected ${executablePath}`);
  }

  if (target === 'client') {
    childProcess.spawnSync('open', [appPath], { stdio: 'inherit' });
  }

  return { executablePath, installKind: 'installed' };
}

function installLinux(
  artifact: DownloadedArtifact,
  { format }: { format: Exclude<ArtifactFormat, 'auto'> }
): InstalledPane {
  if (format === 'deb') {
    const installer = commandExists('apt') ? 'apt' : 'dpkg';
    const args = installer === 'apt' ? ['install', '-y', artifact.path] : ['-i', artifact.path];
    ensureInstallerSucceeded(childProcess.spawnSync('sudo', [installer, ...args], { stdio: 'inherit' }));
    const executablePath = resolveExistingPanePath();
    if (!executablePath) {
      throw new Error('Pane installed from .deb, but the pane executable could not be found.');
    }
    return { executablePath, installKind: 'installed' };
  }

  const binRoot = defaultInstallRoot();
  const executablePath = path.join(binRoot, 'pane');
  fs.mkdirSync(binRoot, { recursive: true });
  fs.copyFileSync(artifact.path, executablePath);
  fs.chmodSync(executablePath, 0o755);
  return { executablePath, installKind: 'installed' };
}

function installWindows(
  artifact: DownloadedArtifact,
  { target }: { target: InstallTarget }
): InstalledPane {
  const args = target === 'daemon' ? ['/S'] : [];
  const result = childProcess.spawnSync(artifact.path, args, { stdio: 'inherit' });
  ensureInstallerSucceeded(result);

  const executablePath = resolveExistingPanePath();
  if (!executablePath) {
    throw new Error('Pane installer completed, but Pane.exe could not be found. Open the installer manually and rerun with --pane-path.');
  }
  return { executablePath, installKind: target === 'daemon' ? 'installed' : 'launched-installer' };
}

function ensureInstallerSucceeded(result: childProcess.SpawnSyncReturns<Buffer>): void {
  if (result.error) throw result.error;
  if (result.status !== 0) {
    const reason = result.signal ? `signal ${result.signal}` : `status ${result.status}`;
    throw new Error(`Pane installer exited with ${reason}.`);
  }
}

function commandExists(command: string): boolean {
  const check = process.platform === 'win32' ? 'where' : 'command';
  const args = process.platform === 'win32' ? [command] : ['-v', command];
  const result = childProcess.spawnSync(check, args, { stdio: 'ignore', shell: process.platform !== 'win32' });
  return result.status === 0;
}

File: packages/runpane-py/src/runpane/installers.py
from __future__ import annotations

import os
import platform
import shutil
import subprocess
from dataclasses import dataclass
from typing import List, Optional

from .download import DownloadedArtifact
from .platforms import PanePlatform, default_install_root


@dataclass
class InstalledPane:
    executable_path: str
    install_kind: str


def resolve_existing_pane_path(pane_path: Optional[str] = None) -> Optional[str]:
    if pane_path:
        return pane_path if os.path.exists(pane_path) else None

    home = os.path.expanduser("~")
    candidates = []
    if os.name == "nt":
        local = os.environ.get("LOCALAPPDATA")
        program_files = os.environ.get("ProgramFiles")
        if local:
            candidates.extend([
                os.path.join(local, "Programs", "Pane", "Pane.exe"),
                os.path.join(local, "Pane", "Pane.exe"),
            ])
        if program_files:
            candidates.append(os.path.join(program_files, "Pane", "Pane.exe"))
    elif platform.system().lower() == "darwin":
        candidates.extend([
            "/Applications/Pane.app/Contents/MacOS/Pane",
            os.path.join(home, "Applications", "Pane.app", "Contents", "MacOS", "Pane"),
        ])
    else:
        candidates.extend([
            os.path.join(home, ".local", "bin", "pane"),
            "/usr/bin/pane",
            "/opt/Pane/pane",
            "/opt/Pane/Pane",
        ])

    for candidate in candidates:
        if os.path.exists(candidate):
            return candidate
    return None


def should_reuse_existing_pane(parsed, target: str) -> bool:
    return parsed.command == "install" and target == "daemon"


def install_pane_artifact(
    artifact: DownloadedArtifact,
    parsed,
    platform: PanePlatform,
    format_name: str,
    target: str,
) -> InstalledPane:
    existing = resolve_existing_pane_path(parsed.pane_path)
    if existing and should_reuse_existing_pane(parsed, target):
        return InstalledPane(executable_path=existing, install_kind="existing")

    if platform.os == "darwin":
        return install_mac(artifact, format_name, target)
    if platform.os == "linux":
        return install_linux(artifact, format_name)
    return install_windows(artifact, target)


def spawn_pane(executable_path: str, args: List[str]) -> int:
    try:
        return subprocess.call(
            [executable_path, *build_pane_daemon_args(args)],
            env=build_pane_daemon_environment(),
        )
    except OSError as error:
        print(f"Failed to launch Pane: {error}")
        return 1


def spawn_pane_captured(executable_path: str, args: List[str]):
    try:
        return subprocess.run(
            [executable_path, *build_pane_daemon_args(args)],
            env=build_pane_daemon_environment(),
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            check=False,
        )
    except OSError as error:
        return subprocess.CompletedProcess([executable_path, *args], 1, "", str(error))


def build_pane_daemon_args(args: List[str], system_name: Optional[str] = None) -> List[str]:
    """Ozone resolves its platform before the app's main script runs, so headless
    selection must arrive as argv. ELECTRON_OZONE_PLATFORM_HINT (below) was removed
    in Electron 38 and is a no-op from 39 on; it is kept only so older Pane builds
    keep working."""
    if (system_name or platform.system()).lower() != "linux":
        return list(args)
    return ["--ozone-platform=headless", "--disable-gpu", *args]


def build_pane_daemon_environment(
    system_name: Optional[str] = None,
    base_environment: Optional[dict] = None,
) -> dict:
    environment = dict(os.environ if base_environment is None else base_environment)
    if (system_name or platform.system()).lower() == "linux":
        environment["ELECTRON_OZONE_PLATFORM_HINT"] = "headless"
    return environment


def launch_pane_client(executable_path: str) -> None:
    subprocess.Popen([executable_path], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, start_new_session=True)


def install_mac(artifact: DownloadedArtifact, format_name: str, target: str) -> InstalledPane:
    if format_name == "dmg":
        subprocess.call(["open", artifact.path])
        return InstalledPane(executable_path="/Applications/Pane.app/Contents/MacOS/Pane", install_kind="launched-installer")

    apps_root = os.path.join(os.path.expanduser("~"), "Applications")
    app_path = os.path.join(apps_root, "Pane.app")
    os.makedirs(apps_root, exist_ok=True)
    subprocess.check_call(["ditto", "-x", "-k", artifact.path, apps_root])
    executable_path = os.path.join(app_path, "Contents", "MacOS", "Pane")
    if not os.path.exists(executable_path):
        raise RuntimeError(f"Pane executable was not found after extracting {artifact.file_name}. Expected {executable_path}")
    if target == "client":
        subprocess.call(["open", app_path])
    return InstalledPane(executable_path=executable_path, install_kind="installed")


def install_linux(artifact: DownloadedArtifact, format_name: str) -> InstalledPane:
    if format_name == "deb":
        installer = "apt" if shutil.which("apt") else "dpkg"
        args = ["install", "-y", artifact.path] if installer == "apt" else ["-i", artifact.path]
        status = subprocess.call(["sudo", installer, *args])
        if status != 0:
            raise RuntimeError(f"Pane installer exited with status {status}.")
        executable = resolve_existing_pane_path()
        if not executable:
            raise RuntimeError("Pane installed from .deb, but the pane executable could not be found.")
        return InstalledPane(executable_path=executable, install_kind="installed")

    bin_root = default_install_root()
    os.makedirs(bin_root, exist_ok=True)
    executable_path = os.path.join(bin_root, "pane")
    shutil.copyfile(artifact.path, executable_path)
    os.chmod(executable_path, 0o755)
    return InstalledPane(executable_path=executable_path, install_kind="installed")


def install_windows(artifact: DownloadedArtifact, target: str) -> InstalledPane:
    args = ["/S"] if target == "daemon" else []
    status = subprocess.call([artifact.path, *args])
    if status != 0:
        raise RuntimeError(f"Pane installer exited with status {status}.")
    executable = resolve_existing_pane_path()
    if not executable:
        raise RuntimeError("Pane installer completed, but Pane.exe could not be found. Open the installer manually and rerun with --pane-path.")
    return InstalledPane(executable_path=executable, install_kind="installed" if target == "daemon" else "launched-installer")
